Cookbook/Technical-Documents/BuildTime-Spikes-Report.md

```diff
@@ -0,0 +1,190 @@
+# Improving our Build Times
+
+Build Times in Swift are known to be quite slow, but we could still try to find ways to reduce them. So far we've tried the following different tactics to reduce our build times especially on CI and clean builds:
+
+ - [➥](#helping-the-type-checker)
+   Rewrite parts of our code differently to help the type-checker  
+   [ [CNSMR-1313](https://babylonpartners.atlassian.net/browse/CNSMR-1313)
+   / [CNSMR-2595](https://babylonpartners.atlassian.net/browse/CNSMR-2595)
+   / [CNSMR-2604](https://babylonpartners.atlassian.net/browse/CNSMR-2604)
+   / [CNSMR-2608](https://babylonpartners.atlassian.net/browse/CNSMR-2608) ]
+ - [➥](#removing-useless-downloads-from-ci)
+   Ensure our CI jobs didn't spend too much time downloading useless things (e.g. `cocoapods-master`)  
+   [ [CNSMR-1517](https://babylonpartners.atlassian.net/browse/CNSMR-1517) ]
+ - [➥](#migrating-to-cocoapods-binary)
+   Investigate the possibility of migrating to `cocoapods-binary` to use pre-compiled Pods on CI  
+   [ [CNSMR-484](https://babylonpartners.atlassian.net/browse/CNSMR-484)
+   / [CNSMR-1513](https://babylonpartners.atlassian.net/browse/CNSMR-1513) ]
+ - [➥](#try-caching-the-deriveddata-for-pods)
+   Try to cache the products from `DerivedData` between runs for tests run on feature PRs  
+   [ [CNSMR-2847](https://babylonpartners.atlassian.net/browse/CNSMR-2847) ]
+
+
+
+## Helping the type-checker
+
+We've installed two Xcode warnings on all our projects to warn us if the type-checker took:
+
+ - More than 150ms on type-checking single expressions (`-Xfrontend -warn-long-expression-type-checking=150`)
+ - More than 200ms on type-checking entire functions (`-Xfrontend -warn-long-function-bodies=200`
+
+We found a lot of those across the codebase. Most common causes were mainly:
+
+* **Long chains of expressions**, due to our heavily-functional codebase, like `x.map(...).flatMap(...).filter(...)` and similar. Those were in fact essentially coming from:
+  * our use of `Bento` to render screens by composing various rows via `|---+` and `|-+` operators
+  * our use of ReactiveSwift/ReactiveCocoa which tend to create long chains like those
+* **Custom operators, especially heavily-overloaded ones and ones that rely on generics**
+  * Especially `Bento` and `Functional.swift` declare a lot of operators that could lead to these usages at point of use, when we have expressions composed with those operators
+  * But also uses of `+` in many places (yes, just `+`). For example `["a","b","c"].joined()` compiles way faster than `"a" + "b" + "c"`. This is probably both caused by `+` being heavily overloaded, by generic implementations nonetheless, but also by the existence of types conforming to `ExpressibleByStringLiteral` that become candidates that the type-checker has to consider during its analysis in these kind of expressions
```

Comment:
this seems to be more like general advices on how writing code affects compile times and how we agree to do that in the team

Reply:
That's also exactly one of the point of the report. While reporting on what we tried to reduce the build time, here's what worked and here is what didn't work, so now that this report is on the playground people can take advantage of that experience and feedback to know what stuff to try rather than others even in future code.

Reply:
But it is hard to extract these bits of something that worked from things that didn't when it all in one huge doc :/

Reply:
I'd say things that didn't work could be restructured into FAQ like `did we try cocoapods-binary?` with short results in form of bullet points of what was the outcome. If anyone would be interested in more in depth explanation we can referent them to some other place, blog post, wiki, whatever
🤷‍♂